### PR TITLE
[OOP][IIachimov] Added Task OOP

### PR DESCRIPTION
**Pull request by IIachimov**
Please, review the submitted code and merge it to the main branch.

### CODE REVIEW

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.List;
+import java.util.UUID;
+
+public abstract class AbstractOrder implements Order {
+  private final String id;
+  private final LocalDateTime date;
+  private List<Item> items;
+
+  public AbstractOrder(List<Item> items) {
+    this.id = id();
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: I would like to recomment you the following: this.id = UUID.randomUUID();
and remove the method id();

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  @Override
+  public int count() {
+    return this.items.size();
+  }
+
+  public int count(Item item) {
+    int totalCountOfItem = 0;
+    for (Item uniqueItem : this.items) {
+      if (uniqueItem.getTitle().equals(item.getTitle())) {
+        totalCountOfItem += 1;
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: replace with totalCountOfItem++;

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.UUID;
+
+public abstract class AbstractUser {
+    private final String id;
+    private final String username;
+    private Map<String, Address> addresses;
+
+
+    public AbstractUser(String username, Map<String, Address> addresses) {
+        this.id = id();
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java]: the same suggestion as above

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AdminUser.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package dataCollectionTask.Models;
+
+import java.util.Map;
+
+public class AdminUser extends AbstractUser {
+    public AdminUser(String username) {
+        super(username);
+    }
+
+    public AdminUser(String username, Map<String, Address> addresses) {
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AdminUser.java]: As far as I know, we should have only one constructor for this class

File: iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java [hunk 5]
```diff
@@ -1,19 +1,50 @@
...
-  public Type type;
+  private final long id;
+  private String title;
+  private Type type;
+  private double price;
   private static long idCounter;
 
-  public Item(String title, Type type) {
+  public Item(String title, Type type, double price) {
     this.id = ++idCounter;
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java]: let's have the same pattern for all ids, let's use UUID as in examples above

File: iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java [hunk 6]
```diff
@@ -1,19 +1,50 @@
...
+
+  public void setType(Type type) {
+    this.type = type;
+  }
+
+  public double getPrice() {
+    return price;
+  }
+
+  public void setPrice(double price) {
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java]: encapsulation is used in order to get the totla control over the variables, can we have a negative price? Could you please handle this situation within the setPrice() method?

File: iulianIachimov/src/main/java/dataCollectionTask/Models/User.java [hunk 7]
```diff
@@ -1,44 +1,14 @@
...
-import java.util.ArrayList;
-import java.util.List;
 import java.util.Map;
 
-public class User {
-  public long id;
-  public String username;
-  public Map<String, Address> addresses;
-  private static long idCounter;
+public class User extends AbstractUser {
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/User.java:5]: where is the User attribute? What is the difference between this class and the parent one>?

File: iulianIachimov/src/main/java/dataCollectionTask/Models/User.java [hunk 8]
```diff
@@ -1,44 +1,14 @@
...
-            + ", username='"
-            + username
-            + '\''
-            + ", addresses="
-            + addresses
-            + '}';
-  }
+    public User(String username) {
+        super(username);
+    }
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/User.java:10]: you should define getters and setters for the User variable

File: iulianIachimov/src/main/java/dataCollectionTask/Models/UserOrder.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+    public UserOrder(List<Item> items, User user) {
+        super(items);
+        this.user = user;
+    }
+
+    public User getUser() {
+        return user;
+    }
+
```
Comment (asecu) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/UserOrder.java]: remove empty lines please

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AdminUser.java (hunk 4 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AdminUser.java]: Per conditions, we had to define a constructor with 1 parameter for class `User`. However, I had already a constructor there with multiple parameters. The common attributes of 'User' class and 'AdminUser' class were moved to 'AbstractUser' and when inherited in subclasses I let both constructors be available. Please let me know if that is the right way to proceed and I can leave the code as it is.

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java (hunk 1 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: Thank you for recommendation, it was implemented.

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java (hunk 2 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: Done

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java (hunk 3 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java]: Done

File: iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java (hunk 5 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java]: Done

File: iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java (hunk 6 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java]: Done

Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/Item.java]: The price should be positive always.

File: iulianIachimov/src/main/java/dataCollectionTask/Models/UserOrder.java (hunk 9 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/UserOrder.java]: Done

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
+package dataCollectionTask.Models;
```
Comment (gaenache) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: According to Java code convention packages should always be written in all-lowercase.
See https://www.oracle.com/technetwork/java/codeconventions-135099.html

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+public abstract class AbstractOrder implements Order {
+  private final UUID id;
+  private final LocalDateTime date;
+  private List<Item> items;
```
Comment (gaenache) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: Please change private to protected

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+public abstract class AbstractUser {
+    private final UUID id;
+    private final String username;
+    private Map<String, Address> addresses;
```
Comment (gaenache) [inline iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java]: Please change private to protected

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java (hunk 10 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: That was changed. However, it caused the previous files to be deleted and new files are committed now.

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java (hunk 11 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractOrder.java]: I am not sure if it is the case to do this move as per conditions we have - Make all fields `private`. I suppose it is a subject to be discussed tomorrow at Q&A Session

File: iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java (hunk 12 above)
Comment (IIachimov) [reply iulianIachimov/src/main/java/dataCollectionTask/Models/AbstractUser.java]: The same as above

**Review by abadarau — APPROVED** [review]

**Review by asecu — APPROVED** [review]

**Review by dvmunteanu — APPROVED** [review]

**Review by constantin73 — APPROVED** [review]

**Pull request merged**